invokeai/app/services/download/download_base.py

```diff
@@ -0,0 +1,200 @@
+    # them from being serialized and/or used in the Json Schema
+    @property
+    def cancelled(self) -> bool:
+        """Call to cancel the job."""
+        return self._cancelled
+
+    @property
+    def on_start(self) -> Optional[DownloadEventHandler]:
+        """Return the on_start event handler."""
+        return self._on_start
+
+    @property
+    def on_progress(self) -> Optional[DownloadEventHandler]:
+        """Return the on_progress event handler."""
+        return self._on_progress
+
+    @property
+    def on_complete(self) -> Optional[DownloadEventHandler]:
+        """Return the on_complete event handler."""
+        return self._on_complete
+
+    @property
+    def on_error(self) -> Optional[DownloadEventHandler]:
+        """Return the on_error event handler."""
+        return self._on_error
+
+    def set_callbacks(
+        self,
+        on_start: Optional[DownloadEventHandler] = None,
+        on_progress: Optional[DownloadEventHandler] = None,
+        on_complete: Optional[DownloadEventHandler] = None,
+        on_error: Optional[DownloadEventHandler] = None,
+    ) -> None:
+        """Set the callbacks for download events."""
+        self._on_start = on_start
+        self._on_progress = on_progress
+        self._on_complete = on_complete
+        self._on_error = on_error
+
+
+class DownloadQueueServiceBase(ABC):
+    """Multithreaded queue for downloading models via URL."""
+
+    @abstractmethod
+    def download(
+        self,
+        source: Union[str, AnyHttpUrl],
+        dest: Union[str, Path],
+        priority: int = 10,
+        access_token: Optional[str] = None,
+        on_start: Optional[DownloadEventHandler] = None,
+        on_progress: Optional[DownloadEventHandler] = None,
+        on_complete: Optional[DownloadEventHandler] = None,
+        on_error: Optional[DownloadEventHandler] = None,
+    ) -> DownloadJob:
+        """
+        Create a download job.
+
+        :param source: Source of the download as a URL.
+        :param destdir: Path to download to. See below.
+        :param on_start, on_progress, on_complete, on_error: Callbacks for the indicated
+         events.
+        :returns: A DownloadJob object for monitoring the state of the download.
+
+        The `dest` argument is a Path object. Its behavior is:
+
+        1. If the path exists and is a directory, then the URL contents will be downloaded
+            into that directory using the filename indicated in the response's `Content-Disposition` field.
+            If no content-disposition is present, then the last component of the URL will be used (similar to
+            wget's behavior).
+        2. If the path does not exist, then it is taken as the name of a new file to create with the downloaded
+           content.
+        3. If the path exists and is an existing file, then the downloader will try to resume the download from
+           the end of the existing file.
```

Comment:
I accidentally created a situation for myself. I downloaded different files to the same destination. The second file was larger than the first, and the download happily "resumed" downloading, resulting in what I imagine would be a nightmare LoRA.

I suppose this will be something we can mostly protect against by just not doing it wrong.

It raises the question, though, if handling partial downloads is something worth supporting (or implementing in the first go). I think the right way to do this is persisting some data about the in-progress downloads to disk. Then we can be confident about resuming downloads, and probably also clean up after ourselves a bit better.

I'll note that my web browser (firefox) does *not* attempt to resume interrupted downloads. It just cancels them if you quit. This seems pretty reasonable to me, as much as I like the idea of resuming partial downloads. Edit: The more I think about it, the more partial downloads sounds like a dark and twisted path, fraught with perils.

Reply:
The resume functionality has been removed. It wasn't going to be much use anyway, as the installer will download its files into a temporary directory that is deleted once finished.